File: solutions/0121. Best Time to Buy and Sell Stock/0121.cs
// Time complexity: O(n)
// Space complexity: O(1)
public class Solution {
    public int MaxProfit(int[] prices) {
        int max = 0;
        int min = prices[0];

        foreach(int price in prices) {
            if(price < min) {
                min = price;
            }
            max = Math.Max(max, price - min);
        }
        return max;
    }
}

File: solutions/0121. Best Time to Buy and Sell Stock/0121.cpp
// Time complexity: O(n)
// Space complexity: O(1)
class Solution {
public:
    int maxProfit(const vector<int>& prices) {
        int ans = 0;
        int prevMin = prices[0];
        
        for (int price : prices) {
            ans = max(ans, price - prevMin);
            prevMin = min(prevMin, price);
        }
        return ans; 
    }
};